common/ui/dialog.cpp

```diff
@@ -194,6 +193,12 @@ void ui_dialog_set_current_canvas(UI_DIALOG *dlg)
 	gr_set_current_canvas(window_get_canvas(*dlg->wind));
 }
 
+UI_DIALOG::~UI_DIALOG()
+{
+	ui_gadget_delete_all(this);
+	selected_gadget = NULL;
```

Comment:
There is no need to update member variables in the destructor.  They go out of scope when the destructor ends, which in this case happens before anything else runs.